text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,557 @@
+would otherwise remain encapsulated. Tracing is also distributed, and requires 
+non-local, transaction-level context propagation in order to execute correctly.
+
+The various observability APIs are not described here directly. However, in this new 
+design, all observability APIs would be modified to make use of the generalized 
+context propagation mechanism described below, rather than the tracing-specific 
+propagation system it uses today.
+
+Note that OpenTelemetry APIs calls should *always* be given access to the entire 
+context object, and never just a subset of the context, such as the value in a 
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+## Correlations API
+
+In addition to trace propagation, OpenTelemetry provides a simple mechanism for 
+propagating metric labels, called the Correlations API. Correlations are 
+intended for indexing events in one service with attributes provided by a prior 
+service in the same transaction. This helps to establish a causal relationship 
+between these events. For example, determining that a particular browser version 
+is associated with a failure in an image processing service.
+
+Correlations are implemented as an extremely basic cross-cutting concern - a 
+transaction-level dictionary. All key-value pairs are stored in the context 
+object, and then injected by the Correlation propagator to be sent to the next 
+process. 
+
+The Correlations API is based on the [W3C Correlation-Context specification](https://w3c.github.io/correlation-context/), 
+and implements the protocol as it is defined in that working group.
+
+While Correlations can be used to prototype other cross-cutting concerns, this 
+mechanism is primarily intended to convey values for the OpenTelemetry 
+observability systems. New concerns with different criteria should be modeled 
+separately, using the same underlying context propagation layer as building 
+blocks.
+
+For backwards compatibility, OpenTracing Baggage is propagated as Correlations 
+when using the OpenTracing bridge.
+
+As every correlation adds additional overhead to every request, correlations 
+should be used with care. The implementation MUST propagate these values. It is 
+entirely on the user to choose when to drop correlations, either by removing 
+them via the API, or by filtering them via a network proxy or service mesh.
+
+**`GetCorrelation(context, key) -> value`**  
+To access the value for a label set by a prior event, the Correlations API 
+provides a function which takes a context and a key as input, and returns a value.
+
+**`SetCorrelation(context, key, value) -> context`**  
+To record the value for a label, the Correlations API provides a function which 
+takes a context, a key, and a value as input, and returns an updated context 
+which contains the new value.
```

Comment:
can you comment on the new value. Is it intended to be a builder pattern and the same, but updated context needs to be reutrned? In other words, people keeping ref on original context - will they see a new value?

Reply:
They would not see the new value.

Here is an example:
https://github.com/tedsuo/rfcs/blob/context-prop-2/text/0066-separate-context-propagation.md#the-scope-of-current-context

BTW, our APIs do a funny thing - hold a reference to a mutable object inside of an immutable context. But in the spec I would prefer to just say immutable. If your language wants to treat the contents of the context as mutable in some cases, then of course this would work differently, and you would need a mutex in order to be thread safe.